docs/features/compilation-from-portable-pdb.md

```diff
@@ -0,0 +1,341 @@
+## compilation-from-portable-pdb
+
+Compilation from portable PDBs today is not completely possible, but is desirable in order to help reconstruct a compilation from source provided via source link or embedded in a pdb. Motivation is derived from [roslyn 41395](https://github.com/dotnet/roslyn/issues/41395).
+
+This document is restricted to the following assumptions:
+
+1. The benefit is for builds with `-deterministic` and published to the symbol server.
+2. Source generator and analyzer references are not needed for this task. They may be useful, but are out of scope for this document.
+3. Any storage capacity used for PDBs and source should not impact this feature, such as compression algorithm.
+4. Only Portable PDB files will be included for this spec. We can expand this feature past these once it is implemented and proven needed elsewhere.
+
+This document will provide the expanded specification to the Portable PDB format. Any additions to that format will be ported to expand documentation provided in [dotnet-runtime](https://github.com/jnm2/dotnet-runtime/blob/26efe3467741fe2a85780b2d2cd18875af6ebd98/docs/design/specs/PortablePdb-Metadata.md#source-link-c-and-vb-compilers).
+
+## List of Compiler Flags
+
+#### CSharp Flags That Can Be Derived From PDB or Assembly
+
+* baseaddress
+* checksumalgorithm
+* debug
+* deterministic
+* embed
+* filealign
+* highentropyva
+* link
+    - Represented by a metadata reference with `EmbededInteropTypes=true` 
+* linkresource
+    - Will be represented in metadata reference embedded in pdb
+* main
+    - Already stored in PDB as the entry point token
+* moduleassemblyname
+* modulename
+* nowin32manifest
+* pdb
+* platform
+* publicsign
+* resource
+    - Will be represented in metadata reference embedded in pdb
+* subsystemversion
+* target
+* win32icon
+* win32manifest
+* win32res
+
+#### CSharp Flags Not Included
+
+* appconfig
+* bugreport
+* delaysign
+* doc
+* errorendlocation
+* errorlog
+* errorreport
+* fullpaths
+* incremental
+* keycontainer
+* keyfile
```

Comment:
All of the key file options meaningfully impact compilation output: delaysign, keyfile and keycontainer. Why are they not being included?

Reply:
If something is signed we can't store the private keys anywhere, and public keys are already available. Maybe worth calling out in this documentation, but I don't know how we would use this information

Reply:
One of the goals here is to be able to accurately re-produce PE files. Given that our recommendation to library authors is still to strong name sign their PE files not having this seems like a problematic gap.

Reply:
We can 100% reproduce strong name - the public key is stored in AssemblyDef metadata table.
Reproducing signature without private key is impossible. The resulting binary will differ in authenticode signature and strong name signature (they will be empty). You can however sign the binary with the private key if you have it and the result will be exactly the same binary.

Reply:
We definitely need tests for various signing cases.

Reply:
I think the tool that re-creates the compilation for validation can set CompilationOptions.PublicSign = true if the COR flags in the DLL have StrongNameSignature bit set, otherwise it would use no signing. That should produce the same binary modulo the actual signatures, correct?

Reply:
> That should produce the same binary modulo the actual signatures, correct?

That matches my understanding.

Reply:
Thought a bit more about this last night. I think it's okay to deliberately omit the key sign data. The only interesting piece of information is the actual public key which is in the assembly. 

The goal here is to be able to verify, by recreating, that this is the correct binary for the given inputs. When it comes to strong names we can simply take the binary we're verifying and strip out the strong name info. Essentially undo strong name and put the binary back into a `publicSign` state. At that point the information in the PDB should be enough to recreate a matching binary (byte for byte)

Reply:
Yup, exactly.